components/eamxx/src/share/field/field_manager.cpp

```diff
@@ -117,109 +129,146 @@ add_to_group (const std::string& field_name, const std::string& group_name)
   if (group.m_info->m_bundled) {
-    if (allocate_subfields) {
-      // We can get the parent from any of the fields in the group.
-      auto p = group.m_fields.begin()->second->get_header().get_parent().lock();
-      EKAT_REQUIRE_MSG(p!=nullptr,
-          "Error! A field belonging to a bundled field group is missing its 'parent'.\n");
-
-      const auto& p_id = p->get_identifier();
-      group.m_bundle = get_field_ptr(p_id);
+    // All fields in a group have the same parent, get the parent from the 1st one
+    const auto& parent_header = group.m_fields.begin()->second->get_header().get_parent().lock();
+    const auto& parent_id = parent_header->get_identifier();
+    const auto& parent_field = get_field_ptr(parent_id);
+
+    EKAT_REQUIRE_MSG(parent_header!=nullptr,
+      "Error! A field belonging to a bundled field group is missing its 'parent'.\n");
+
+    // Get the parent field of one of the group fields and check if
+    // all it's children are in the group.
+    bool all_children_in_group = true;
+    for (auto child : parent_header->get_children()) {
+      bool in_group = false;
+      for (auto f : group.m_fields) {
+        if (child.lock()->get_identifier() == f.second->get_header().get_identifier()) {
+          in_group = true;
+          break;
+        }
+      }
+      if (not in_group) {
+        all_children_in_group = false;
+        break;
+      }
+    }
+
+    // Set m_bundle field
+    if (all_children_in_group) {
+      // If all children are in the group, this is the parent group and
+      // we can just query the bundle and set here
+      group.m_bundle = parent_field;
     } else {
-      // This group was copied, and therefore we know we created the bundled field with
-      // the same name as the group itself. So simply fetch that field
-      group.m_bundle = get_field_ptr(group_name);
+      // If children exist that aren't in this group, we need to get a
+      // subfield of the parent group containing indices of group fields
+      std::vector<int> ordered_subview_indices;
+      for (auto fn : group.m_info->m_fields_names) {
+        const auto parent_child_index =
+          m_field_groups.at(parent_field->name())->m_subview_idx.at(fn);
+        ordered_subview_indices.push_back(parent_child_index);
+      }
+      std::sort(ordered_subview_indices.begin(),ordered_subview_indices.end());
+
+      // Check that all subview indices are contiguous
+      for (size_t i=0; i<ordered_subview_indices.size()-1; ++i) {
+        EKAT_REQUIRE_MSG(ordered_subview_indices[i]+1==ordered_subview_indices[i+1],
+          "Error! Non-contiguous subview indices found in group \""+group_name+"\"\n");
+      }
+
+      group.m_bundle = std::make_shared<Field>(
+        parent_field->subfield(
+          group_name, parent_header->get_identifier().get_units(),
+          group.m_info->m_subview_dim,
+          ordered_subview_indices.front(), ordered_subview_indices.back()));
```

Comment:
In C++ is more common to express ranges with the "half-open" convention, where the right end point is not included. That's what begin/end iterators are in containers (they represent the `[begin,end)` range). So I would encourage our subfield method to behave the same way (so here, you would have to add +1 to the last arg).